### PR TITLE
8293590: Some syntax checks performed by URL.openConnection() could be performed earlier, at URL construction

### PR DESCRIPTION
**Pull request by dfuch**
Many built-in URL Handlers perform additional syntax checking on the URL when `URLStreamHandler::openConnection` / connect is called. In some cases, some of these checks could be also performed earlier, when `URLStreamHandler::parseURL` is called.

This fix proposes to slightly modify the behavior of JDK built-in URL handlers to perform these checks early, and throw `MalformedURLException` earlier, at URL construction time, when that is possible, instead of delaying until `URL::openConnection` or `URLConnection::connect` are called.

Because java.net.URL constructors are already specified to throw `MalformedURLException` in such cases, no API documentation changes are necessary. 

A new JDK specific system property `-Djdk.net.url.delayParsing` or `-Djdk.net.url.delayParsing=true` can be specified on the command line to revert to the previous behavior. By default, the property is not set, and the new behavior is in place. The property can be specified on the java command line if unexpected regressions are observed.

The new property will be documented in release notes, and may be removed in a future release.

CSR: https://bugs.openjdk.org/browse/JDK-8294592

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8293590](https://bugs.openjdk.org/browse/JDK-8293590): Some syntax checks performed by URL.openConnection() could be performed earlier, at URL construction
 * [JDK-8294592](https://bugs.openjdk.org/browse/JDK-8294592): Some syntax checks performed by URL.openConnection() could be performed earlier, at URL construction (**CSR**)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [b612f6e4](https://git.openjdk.org/jdk/pull/10534/files/b612f6e4020f6590fe811f426c059dcbdf2393cd)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [d01b898d](https://git.openjdk.org/jdk/pull/10534/files/d01b898d480386a4583f7891405d5ff0ee4cc4b4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10534/head:pull/10534` \
`$ git checkout pull/10534`

Update a local copy of the PR: \
`$ git checkout pull/10534` \
`$ git pull https://git.openjdk.org/jdk pull/10534/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10534`

View PR using the GUI difftool: \
`$ git pr show -t 10534`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10534.diff">https://git.openjdk.org/jdk/pull/10534.diff</a>

</details>

### CODE REVIEW

**Review by jaikiran — APPROVED** [review]
The changes look fine to me.

File: src/java.base/share/classes/java/net/URLStreamHandler.java [hunk 1]
```diff
@@ -504,10 +504,23 @@ protected void setURL(URL u, String protocol, String host, int port,
...
-            if (s != null) throw new IllegalArgumentException(s);
+                    "this handler");
+        }
+        // if early parsing, perform additional checks here rather than waiting
+        // for openConnection()
+        boolean earlyURLParsing = IPAddressUtil.earlyURLParsing();
+        boolean isBuiltInHandler = u.isBuiltinStreamHandler(this);
+        if (host != null && isBuiltInHandler) {
+            String errMsg = IPAddressUtil.checkHostString(host);
+            if (errMsg != null) throw new IllegalArgumentException(errMsg);
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/net/URLStreamHandler.java:515]: Is it time for URLStreamHandler.setURL to document the IllegalArgumentException?

File: src/java.base/share/classes/sun/net/util/IPAddressUtil.java [hunk 2]
```diff
@@ -479,10 +489,14 @@ private static String describeChar(char c) {
         return "'" + c + "'";
     }
 
-    private static String checkUserInfo(String str) {
+    // Check user-info component.
+    // This method returns an error message if a problem
+    // is found. The caller is expected to use that message to
+    // throw an exception.
+    public static String checkUserInfo(String str) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/net/util/IPAddressUtil.java:496]: I think IPAddressUtil was original introduced as a utility class to support parsing of IP addresses and do text <--> byte[] conversion. It has grown into checking the authority and userInfo components so I think is begging for a rename or some refactoring. It's not this PR but I hope you see what I mean on the naming/purpose of this class.

**Comment by AlanBateman** [comment]
At a high-level I think this proposal is okay and rejecting nonsense URLs early is the right thing to do. Having MalformedURLException thrown if the "parsed URL fails to comply with the specific syntax of the associated protocol" already sets expectations that the input will be rejected at URL construction time.

As regards the compatibility impact then it should be low. Code that creates a bad URL and opens a connection may not notice the change, except maybe that an exception is thrown in a different place. Code that creates bad URLs but doesn't attempt to open connections may get an exception when they didn't previously but it should be rare and there is a system property to work around the behavior change until the code is fixed to use the correct URL syntax. So overall I think this is okay.

File: src/java.base/share/classes/java/net/URLStreamHandler.java (hunk 1 above)
Comment (dfuch) [reply src/java.base/share/classes/java/net/URLStreamHandler.java:515]: Good point. I have logged https://bugs.openjdk.org/browse/JDK-8294948 to follow-up on this.

File: src/java.base/share/classes/sun/net/util/IPAddressUtil.java (hunk 2 above)
Comment (dfuch) [reply src/java.base/share/classes/sun/net/util/IPAddressUtil.java:496]: Yes - I see. I have logged https://bugs.openjdk.org/browse/JDK-8294951

**Comment by dfuch** [comment]
Thanks @AlanBateman ! Do I have your approval then?

**Review by Michael-Mc-Mahon — APPROVED** [review]
I note that the change to throw `MalformedURLException` instead of IAE in FtpURLConnection is not covered by the compatibility system property, but the probability of actual compatibility issues with that seems pretty remote, considering IOE already has to be handled. Change looks fine to me.

**Comment by AlanBateman** [comment]
> Thanks @AlanBateman ! Do I have your approval then?

I've added myself as Reviewer to the CSR.

**Pull request closed without merging**